### ISSUE TITLE
Websocket fails when using authentication

### ISSUE THREAD

**Issue by rleibman** [akka/akka#19689]
I'm having an issue with websockets using authentication:

WebSocket connection to 'wss://www.example.com:8083/processStatusUpdate' failed: Connection closed before receiving a handshake response

My server route looks like this:

      authenticateLdap(authenticator.apply _) {
        implicit user ⇒
          path("processStatusUpdate") {
            handleWebsocketMessages {
              processStatusUpdateFlow(LdapUser(user))
            }
          }
      }

If I remove the auth portion (and just pass a manually created user), this works perfectly.
The authentication portion works everywhere else in the route, and it is Making it inside my processStatusUpdateFlow (as my log shows), so the user does get properly authenticated.

Here's what my authenticateLdap looks like:

```Scala
  def authenticateLdap[T](authenticatorFn: (Option[UserPass] ⇒ Future[Option[T]])): AuthenticationDirective[T] = {
    authenticateOrRejectWithChallenge[BasicHttpCredentials, T] {
      case Some(BasicHttpCredentials(username, password)) ⇒
        val auth = authenticatorFn(Some(UserPass(username, password)))
        auth.map {
          case Some(t) ⇒ AuthenticationResult.success(t)
          case None    ⇒ AuthenticationResult.failWithChallenge(HttpChallenge(scheme = "Basic", realm = "", params = Map.empty))
        }
      case None ⇒ Future.successful(AuthenticationResult.failWithChallenge(HttpChallenge(scheme = "Basic", realm = "", params = Map.empty)))
    }
  }
```

**Comment by rleibman** [comment]
TLDR;
handleWebsocketMessages don't seem to work inside authenticateOrRejectWithChallenge

**Comment by rleibman** [comment]
ok... more information. It seems to have something to do with Futures and running on a different thread:

I've reduced the problem to:

    path("processStatusUpdate") {
      authenticateOrRejectWithChallenge(myauth _) {
        implicit user ⇒
          handleWebsocketMessages {
            processStatusUpdateFlow(user)
          }
      }
    } ~

**This Works**
  def myauth(credentials: Option[HttpCredentials]): Future[AuthenticationResult[LdapUser]] = {
    Future.successful {
      AuthenticationResult.success(LdapUser(dn = "rleibman", uid = "rleibman", givenName = "rleibman", surname = "rleibman", mail = "rleibman"))
    }
  }

**This does NOT work**
  def myauth(credentials: Option[HttpCredentials]): Future[AuthenticationResult[LdapUser]] = {
    Future {
      AuthenticationResult.success(LdapUser(dn = "rleibman", uid = "rleibman", givenName = "rleibman", surname = "rleibman", mail = "rleibman"))
    }
  }

**Comment by johanandren** [comment]
Based on your details I created this: https://gist.github.com/johanandren/964672acc37b84caca40 which works perfectly fine accessed from both Safari and Chrome. There must be something else in your application causing trouble?

**Comment by rleibman** [comment]
Ok... try the same thing with Future instead of Future.successful

**Comment by johanandren** [comment]
Yup, sorry, I was moving back and forth between them and pasted the wrong one. It works with both here. 

Have you done any custom configuration of akka-http in your `application.conf` perhaps?

**Comment by rleibman** [comment]
Thanks for taking a look into this. I doubt this would make it a problem, but here it is.

akka {
	loglevel = "debug"

	http.host-connection-pool {
	 	max-connections = 64
		max-open-requests = 64
	}
}

Mhh... if it works for you, I wonder if it's a timing problem? What if you add a wait on your test? I can play with your test and try to see if I can reproduce it. What do you use as your client?

**Comment by johanandren** [comment]
Tried both Chrome and Safari with the html included in the comment in the sample.

**Comment by johanandren** [comment]
A 2 second delay before returning the future value still works here.

**Comment by johanandren** [comment]
(of course it could be a race condition the other way around, but I don't think that is likely)

**Comment by evolchek** [comment]
Hi, I'm facing the same issue. My code is looking like this:

```
(path("ws" / "agents" / Segment / "metrics") & parameters(("secret", "numberOfLastMetrics".as[Int]))) {
  (agentName, secret, numberOfLastMetrics) =>
    onSuccess(agentRestController.validateAgent(agentName, secret)) { maybeAgent =>
      validate(maybeAgent.nonEmpty, s"Agent is not valid") {
        handleWebsocketMessages(someFlow())
      }
    }
}
```
In my case the `onSuccess` directive causes the failure. Please note that the problem is present in versions 2.0.2 to 2.0.3 of Akka HTTP. Versions 2.0.1 and older work fine.

**Comment by evolchek** [comment]
Oh, I just realised that Akka HTTP is not experimental any more. v2.4.2 works fine. Thanks!

**Comment by ktoso** [comment]
Do I understand correctly that with 2.4.2 you don't see the problem anymore and we can close this ticket?

**Comment by evolchek** [comment]
Right.

**Comment by ktoso** [comment]
Great news, closing the ticket then – thanks for confirming!
